src/query.js

```diff
@@ -1,242 +1,393 @@
    * @typedef {queryFunc} function
-   * @param {PeerId} next - Peer to query
-   * @param {function(Error, Object)} callback - Query result callback
+   * @param {PeerId} peer - Peer to query
+   * @returns {closerPeers: Array<PeerInfo>, value: Object, pathComplete: bool, queryComplete: bool}
    */
 
   /**
-   * Create a new query. The makePath function is called once per disjoint path, so that per-path
-   * variables can be created in that scope. makePath then returns the actual query function (queryFunc) to
+   * Create a new query. The makePathQuery function is called once per disjoint path, so that per-path
+   * variables can be created in that scope. makePathQuery then returns the actual query function (queryFunc) to
    * use when on that path.
    *
    * @param {DHT} dht - DHT instance
    * @param {Buffer} key
-   * @param {makePath} makePath - Called to set up each disjoint path. Must return the query function.
+   * @param {makePathQuery} makePathQuery - Called to set up each disjoint path. Must return the query function.
    */
-  constructor (dht, key, makePath) {
+  constructor (dht, key, makePathQuery) {
     this.dht = dht
     this.key = key
-    this.makePath = makePath
-    this.concurrency = c.ALPHA
+    this.makePathQuery = makePathQuery
     this._log = utils.logger(this.dht.peerInfo.id, 'query:' + mh.toB58String(key))
+    this.running = false
+  }
+
+  /**
+   * Called when the Query starts.
+   */
+  _onStart () {
+    this.running = true
+    this.dht._queryManager.started(this)
+    this._log('run:start')
+  }
+
+  /**
+   * Stop the Query if it is running. This should be called after run() if the
+   * Query hasn't completed (eg if the caller only needs a limited number of
+   * results) so that the Query doesn't continue traversing the DHT
+   * unnecessarily.
+   */
+  stop () {
+    this.run && this.run.stop()
+
+    if (this.running) {
+      this.running = false
+      this._log('run:complete')
+      this.dht._queryManager.stopped(this)
+    }
   }
 
+  /**
+   * Run result.
+   * @typedef {Object} RunResult
+   * @property {Array<PeerId>} peersSeen - peers that have been queried so far
+   * @property {Object} [value] - a value returned in the value field of the queryFunc
+   */
+
   /**
    * Run this query, start with the given list of peers first.
    *
    * @param {Array<PeerId>} peers
-   * @param {function(Error, Object)} callback
-   * @returns {void}
-   */
-  run (peers, callback) {
-    const run = {
-      peersSeen: new Set(),
-      errors: [],
-      paths: null // array of states per disjoint path
-    }
-
+   * @param {number} [timeout] - timeout in ms. If undefined, runs forever.
+   * @returns {AsyncIterator<RunResult>}
+   */
+  async * run (peers, timeout) {
     if (peers.length === 0) {
       this._log.error('Running query with no peers')
-      return callback()
+      return
+    }
+
+    const run = new Run(this.dht, this.key, this.makePathQuery, this._log)
+    this.run = run
```

Comment:
seems like `this.run` accidentally clashes with the `run` function defined on the `Query` class. I guess this could be an intentional override, but if so it should be heavily commented because its surprising

Reply:
@dirkmc could you clarify if this is intentional or not

Reply:
@kumavis it's not intentional - thanks for pointing it out.

I think it's best to hold off on updating this PR until the dependencies are released

Reply:
opened an issue https://github.com/libp2p/js-libp2p-kad-dht/issues/114